First draft of static code analysis (automated code analysis) for rev…

nicolelynn323 opened the pull request:
…iew.

dcmoore commented on practices/perform-static-code-analysis.md:
```diff
@@ -1,115 +1,71 @@
-Review the initial set of issues identified to understand the tool's strengths and weaknesses, and determine which tool aligns best with your workflow.
+- *Automation Should Complement, Not Replace Human Review* – Automated tools are great at spotting patterns but can miss context-specific problems. Keep human judgment in the loop.
+- *False Positives Can Cause Fatigue* – Too many non-issues flagged will erode trust in the tool. Customize rulesets and adjust sensitivity over time.
+- *Makes Tight Coupling Easier* - When you make it easier to write code that changes multiple parts of the broader system, you make it easier to introduce code that doesn't separate concerns and increases module coupling. This can be mitigated with constant retrospectives focused on how these boundaries are formed. The good news about monorepos is when you identify a poorly constructed boundary,  ...
```
Looks like this was copied here by mistake.

dcmoore commented on practices/perform-static-code-analysis.md:
```diff
@@ -1,115 +1,71 @@
+- *Use the Right Tool for the Right Job* – Some tools shine in specific languages or environments; a multi-tool approach often works best.
 
-### [Lead Workshops](/practices/lead-workshops.md)
+## How to Gain Traction
```
Can you add a step 1 that's designed to educate the team members on how to use these sort of tools? Is there a (or multiple) talk or blog post that covers this ground well that we can recommend the team watches before diving in?

dcmoore commented on practices/perform-static-code-analysis.md:
```diff
@@ -1,115 +1,71 @@
-Integrate the chosen tool with your version control system and IDE.
-Review the initial set of issues identified to understand the tool's strengths and weaknesses, and determine which tool aligns best with your workflow.
+- *Automation Should Complement, Not Replace Human Review* – Automated tools are great at spotting patterns but can miss context-specific problems. Keep human judgment in the loop.
+- *False Positives Can Cause Fatigue* – Too many non-issues flagged will erode trust in the tool. Customize rulesets and adjust sensitivity over time.
```
There's another concept here coined by Michael Feathers called [review fatigue](https://medium.com/@sageniuz/where-ai-meets-code-techniques-and-best-practices-from-michael-feathers-a-summary-312ef91b6472). Perhaps that's worth incorporating here?

dcmoore commented on practices/perform-static-code-analysis.md:
```diff
@@ -1,115 +1,71 @@
+- You are a Tech Lead or Manager and you need to ensure consistent code quality across a team without increasing review overhead.
 
-### Importance of Developer Judgment
+## Goals, Metrics & Signals
```
I apologize, but I've moved the goal posts ever so slightly on this section. Can you review the updated [template](https://github.com/pragmint/open-practices/blob/main/templates/new-practice.md) and [example practice](https://github.com/pragmint/open-practices/blob/main/practices/migrate-to-monorepo.md) and make tweaks to follow that new format?

dcmoore replied:
Also, if after reviewing, you have differing opinions, feel free to share those [here](https://github.com/pragmint/open-practices/pull/52) even though the PR is closed.

dcmoore commented:
@ntache-81 let's wait for the next iteration.

ntache-81 commented:
Hi @nicolelynn323 -- I've just committed my edits to the file. It's in GREAT shape. See comments in "Files changed" tab.

ntache-81 commented on practices/perform-static-code-analysis.md:
```diff
@@ -1,115 +1,79 @@
+- Self-hosted LLMs: Tools like Ollama or LM Studio allow you to run open-source AI models locally
 
-### Common Misconceptions about Static Code Analysis
+## Who It’s For & Why
```
In this section, I added "so they can..." language to surface what I understand the individual's goal(s) to be. Please review to ensure it's reading accurately.

ntache-81 commented on practices/perform-static-code-analysis.md:
```diff
@@ -1,115 +1,79 @@
+- **Tech leads or managers** need to enforce consistent code quality across the team so they can deliver successful products without increasing review overhead. 
 
-### Importance of Developer Judgment
+## How to Gain Traction
```
I really like how the items in this section are sequential. That makes it feel super actionable, almost like a checklist.

ntache-81 commented on practices/perform-static-code-analysis.md:
```diff
@@ -1,115 +1,79 @@
-* Are there any bottlenecks or disruptions caused by static code analysis tools in our current workflow?
-
-#### Developer Judgment
+## Supporting Capabilities
```
Nice, snappy one-liners in this section. Great!

ntache-81 commented on practices/perform-static-code-analysis.md:
```diff
@@ -1,115 +1,79 @@
+Choose one active repository and integrate one or two automated analysis tools (both a static analyzer and, optionally, an AI assistant). Measure how quickly developers address flagged issues and collect feedback.
 
-### [Do A Spike](/practices/do-a-spike.md)
+### Optimize Rules and Initiate Feedback Loops
```
Is "Initiate" the right word here? I added it because it doesn't sound like we're "optimizing" the feedback loops as we are the rules; rather, we're generating feedback loops. Right?

ntache-81 commented on practices/perform-static-code-analysis.md:
```diff
@@ -1,115 +1,79 @@
-### [Start A Book Club](/practices/start-a-book-club.md)
+- ...fewer issues are flagged during manual code reviews because they have been automatically detected. Track this by tagging review comments or using tools like GitHub's review insights, [DX](https://getdx.com/platform/data-lake/), or [Code Climate Velocity](https://docs.velocity.codeclimate.com/en/) to analyze trends over time.
+- ...there is a decrease in production bugs linked to preventable errors (e.g., null checks, insecure patterns). Teams can monitor this trend by tagging incident postmortems or using bug categorization in tools like [Jira](https://support.atlassian.com/jira-cloud-administration/docs/what-are-issue-types/), [Linear](https://linear.app/docs/labels), or observability platforms like [Sentry](https:// ...
+- ...developer sentiment around code review “friction” improves. You can capture this through lightweight surveys using [Typeform](https://www.typeform.com/) or [Google Forms](https://www.google.com/forms/about/) before and after adoption. These can be incorporated into team retros; look for signals like reduced frustration with nitpicky feedback or faster review turnaround times.
```
When we say "**These** can be incorporated into team retros," _what_ can be incorporated? Perceived sentiment? Tools to measure sentiment? Small point, but suggest clarifying.

ntache-81 commented on practices/perform-static-code-analysis.md:
```diff
@@ -1,115 +1,79 @@
+### [Code Maintainability](/capabilities/code-maintainability.md)
 
-#### Issue Detection and Resolution
+Automated code analysis enforces consistent coding standards and identifies maintainability issues early, keeping the codebase clean and flexible.
```
Is my edit to this sentence OK (editing "approachable" to "flexible")? I wanted to replace "approachable" with a word that might better convey code that can be easily maintained.

dcmoore commented:
I updated the format to follow the newly suggested practice format from #61. I also revised the "available tools" listing at the top and extracted the resource pages from this practice page. Can the two of you take a look?

dcmoore commented on resources/tech/where-ai-meets-code.md:
```diff
@@ -0,0 +1,39 @@
+
+Resource type: Video
+
+https://www.youtube.com/watch?v=g9m3R0NMJ1Y
```
@nicolelynn323 I wound up using the original video instead of the blog post about the video you originally linked to. I figured this is the more defacto link to reference.

dcmoore commented:
Looking at this closer, it seems to cover the same ground as the [Automate Coding Standards](https://github.com/pragmint/open-practices/blob/main/practices/automate-coding-standards.md) practice. Should we extract any resources / points from that one and load them into this practice, then delete the automate coding standards practice?

nicolelynn323 commented:
> I updated the format to follow the newly suggested practice format from #61. I also revised the "available tools" listing at the top and extracted the resource pages from this practice page. Can the two of you take a look?

I really like the changes you made here, I will make sure to utilize these updated formats for the Otel practice.

ntache-81 commented on practices/perform-automated-code-analysis.md:
```diff
@@ -37,21 +37,17 @@ Assuming the pilot went well, gather the team to share results and best practice
 
 ## Lessons From The Field
```
Removed "Lesson" lines, as our other recent Practices do not have this. Incorporated "Lesson" sentiments into the paragraph.

ntache-81 commented on practices/perform-automated-code-analysis.md:
```diff
@@ -71,14 +67,11 @@ After experimenting with this practice for **4–5 weeks**, bring the team toget
 
 ## How to Improve
```
@dcmoore -- Is this a section we want to include in Practices going forward? Say, in the instance the client needs some tips on how to polish if metrics/signals need improving? I want to be sure I understand the purpose of this section so it can be implemented consistently across Practices.

dcmoore commented on practices/perform-automated-code-analysis.md:
```diff
@@ -13,9 +13,9 @@ Some popular tools for automated code analysis include:
-- **Developers** need fast feedback on bugs, design issues, and inconsistencies so they can work more efficiently and avoid waiting for review cycles.
-- **QA engineers** need to identify high-risk areas earlier so they can effectively focus their limited testing time.
-- **Tech leads or managers** need to enforce consistent code quality across the team so they can deliver successful products without increasing review overhead.
+- "I am a developer and I need fast feedback on bugs, design issues, and inconsistencies so I can work more efficiently and avoid waiting for review cycles."
```
Sorry, I neglected to update the template. I fixed this to be `You are a` and removed the quotes to stay consistent with the other practices.
